examples/research_projects/codeparrot/scripts/codeparrot_training.py

```diff
@@ -0,0 +1,232 @@
+        logger.setLevel(logging.ERROR)
+        datasets.utils.logging.set_verbosity_error()
+        transformers.utils.logging.set_verbosity_error()
+    return logger, tb_writer, run_name
+
+
+def create_dataloaders(dataset_name, args):
+    ds_kwargs = {"streaming": True}
+    train_data = load_dataset(dataset_name + "-train", split="train", **ds_kwargs)
+    train_data = train_data.shuffle(buffer_size=args.shuffle_buffer, seed=args.seed)
+    valid_data = load_dataset(dataset_name + "-valid", split="train", **ds_kwargs)
+    train_dataset = ConstantLengthDataset(tokenizer, train_data, infinite=True, seq_length=args.seq_length)
+    valid_dataset = ConstantLengthDataset(tokenizer, valid_data, infinite=False, seq_length=args.seq_length)
+    train_dataloader = DataLoader(train_dataset, batch_size=args.train_batch_size)
+    eval_dataloader = DataLoader(valid_dataset, batch_size=args.valid_batch_size)
+    return train_dataloader, eval_dataloader
+
+
+def get_grouped_params(model, args, no_decay=["bias", "LayerNorm.weight"]):
+    params_with_wd, params_without_wd = [], []
+    for n, p in model.named_parameters():
+        if any(nd in n for nd in no_decay):
+            params_without_wd.append(p)
+        else:
+            params_with_wd.append(p)
+    return [
+        {"params": params_with_wd, "weight_decay": args.weight_decay},
+        {"params": params_without_wd, "weight_decay": 0.0},
+    ]
+
+
+def log_metrics(step, metrics):
+    logger.info(f"Step {step}: {metrics}")
+    if accelerator.is_main_process:
+        wandb.log(metrics)
+        [tb_writer.add_scalar(k, v, step) for k, v in metrics.items()]
+
+
+def evaluate(args):
+    model.eval()
+    losses = []
+    for step, batch in enumerate(eval_dataloader):
+        with torch.no_grad():
+            outputs = model(batch, labels=batch)
+        loss = outputs.loss.repeat(args.valid_batch_size)
+        losses.append(accelerator.gather(loss))
+        if args.max_eval_steps > 0 and step >= args.max_eval_steps:
+            break
+    loss = torch.mean(torch.cat(losses))
+    try:
+        perplexity = torch.exp(loss)
+    except OverflowError:
+        perplexity = float("inf")
+    return loss.item(), perplexity.item()
+
+
+# Accelerator
+accelerator = Accelerator()
+acc_state = {str(k): str(v) for k, v in accelerator.state.__dict__.items()}
+
+# Hyperparameters (codeparrot-small configs are in comments)
+project_name = "lvwerra/codeparrot"
+dataset_name = "../codeparrot-clean"
+config = {
+    "train_batch_size": 2,  # 16
+    "valid_batch_size": 2,  # 16
+    "weight_decay": 0.1,
+    "shuffle_buffer": 1_000,
+    "learning_rate": 2e-4,  # 5e-4
+    "lr_scheduler_type": "cosine",
+    "num_warmup_steps": 750,  # 2000
+    "gradient_accumulation_steps": 16,  # 1
+    "gradient_checkpointing": True,  # False
+    "max_train_steps": 50_000,  # 150_000
+    "max_eval_steps": -1,
+    "seq_length": 1024,
+    "seed": 1,
+    "save_checkpoint_steps": 50_000,
+}  # 15_000
```

Comment:
Dead code?

Reply:
no, this is the codeparrot-small setting (see comment at beginning of config)

Reply:
Ah, I see. If you plan to keep this comment-based approach, then perhaps the `15_000` needs to be moved within the dict?